### PR TITLE
Add 'make_json' output plugin

### PR DESCRIPTION
**Pull request by metaMMA**
### Motivation for changes: 
I need to run tests on the entry fields. Specifically, I need to take a custom entry field containing a datetime string, convert it to a datetime object, make some replacements, run a test on the object and create new files with the data. Loading a .json file into a python script makes for easy parsing via a python dictionary.

### Detailed changes: 
- Added two new files. 
- The first file creates a `make_json` output plugin that takes all entries in an entry list and saves them to a .json file with structure determined by a jinja template. 
- The second is a default jinja template for formatting the .json file. By default, the template creates a key that is the `title` of the entry paired with an object. That object is all of the the entry's keys paired with their values.

### Config usage if relevant (new plugin or updated schema):
identical to `make_html`

### Log and/or tests output (preferably both):
```
2019-11-06 20:22 INFO     manager                       Test mode, creating a copy from database ...
2019-11-06 20:22 INFO     manager                       Test database created
2019-11-06 20:22 INFO     performance                   Enabling plugin and SQLAlchemy performance debugging
2019-11-06 20:22 VERBOSE  task_queue                    There are 1 tasks to execute. Shutdown will commence when they have completed.
2019-11-06 20:22 VERBOSE  details       calendar_dump   Produced 5 entries.
2019-11-06 20:22 VERBOSE  task          calendar_dump   ACCEPTED: `UFC 246` by accept_all plugin
2019-11-06 20:22 VERBOSE  task          calendar_dump   ACCEPTED: `UFC on ESPN+ 23: Ortega vs. Korean Zombie` by accept_all plugin
2019-11-06 20:22 VERBOSE  task          calendar_dump   ACCEPTED: `UFC 245: Usman vs. Covington` by accept_all plugin
2019-11-06 20:22 VERBOSE  task          calendar_dump   ACCEPTED: `UFC on ESPN 7: Overeem vs. Rozenstruik` by accept_all plugin
2019-11-06 20:22 VERBOSE  task          calendar_dump   ACCEPTED: `UFC on ESPN+ 22: Blachowicz vs. Jacare` by accept_all plugin
2019-11-06 20:22 VERBOSE  details       calendar_dump   Summary - Accepted: 5 (Rejected: 0 Undecided: 0 Failed: 0)
2019-11-06 20:22 VERBOSE  make_json     calendar_dump   Writing output json to ./metaMMA/Flexget/samplejson.json
2019-11-06 20:22 INFO     performance                   Performance results for task calendar_dump:
2019-11-06 20:22 INFO     performance                   status          took 0.24 sec (4 queries)
2019-11-06 20:22 INFO     performance                   remember_rejected took 0.14 sec (4 queries)
2019-11-06 20:22 INFO     performance                   inputs          took 0.75 sec (0 queries)
2019-11-06 20:22 INFO     performance                   retry_failed    took 0.02 sec (25 queries)
2019-11-06 20:22 INFO     manager                       Removed test database

```

JSON file:
```
{ 
"UFC 246": {
"original_title": "UFC 246",
"title": "UFC 246",
"original_url": "<REDACTED_BY_metaMMA>ufc-fight-night",
"url": "<REDACTED_BY_metaMMA>ufc-fight-night",
"datetime": "Saturday, January 18,  5:00 PM ET",
"task": "calendar_dump",
"quality": "unknown",
"accepted_by": "accept_all"
},
"UFC on ESPN+ 23: Ortega vs. Korean Zombie": {
"original_title": "UFC on ESPN+ 23: Ortega vs. Korean Zombie",
"title": "UFC on ESPN+ 23: Ortega vs. Korean Zombie",
"original_url": "<REDACTED_BY_metaMMA>ufc-on-espn-23",
"url": "<REDACTED_BY_metaMMA>ufc-on-espn-23",
"datetime": "Saturday, December 21,  2:00 AM ET",
"task": "calendar_dump",
"quality": "unknown",
"accepted_by": "accept_all"
},
"UFC 245: Usman vs. Covington": {
"original_title": "UFC 245: Usman vs. Covington",
"title": "UFC 245: Usman vs. Covington",
"original_url": "<REDACTED_BY_metaMMA>ufc-245",
"url": "<REDACTED_BY_metaMMA>ufc-245",
"datetime": "Saturday, December 14, 10:00 PM ET",
"task": "calendar_dump",
"quality": "unknown",
"accepted_by": "accept_all"
},
"UFC on ESPN 7: Overeem vs. Rozenstruik": {
"original_title": "UFC on ESPN 7: Overeem vs. Rozenstruik",
"title": "UFC on ESPN 7: Overeem vs. Rozenstruik",
"original_url": "<REDACTED_BY_metaMMA>ufc-on-espn-6",
"url": "<REDACTED_BY_metaMMA>ufc-on-espn-6",
"datetime": "Saturday, December 07,  5:30 PM ET",
"task": "calendar_dump",
"quality": "unknown",
"accepted_by": "accept_all"
},
"UFC on ESPN+ 22: Blachowicz vs. Jacare": {
"original_title": "UFC on ESPN+ 22: Blachowicz vs. Jacare",
"title": "UFC on ESPN+ 22: Blachowicz vs. Jacare",
"original_url": "<REDACTED_BY_metaMMA>ufc-on-espn-21",
"url": "<REDACTED_BY_metaMMA>ufc-on-espn-21",
"datetime": "Saturday, November 16,  5:00 PM ET",
"task": "calendar_dump",
"quality": "unknown",
"accepted_by": "accept_all"
}
}

#### To Do:

- [ ] Discuss which entry keys should be in the default `json.template` file.

### CODE REVIEW

**Comment by gazpachoking** [comment]
Does jinja really get us anything here? Seems like we can just do a json.dump to get a json output.

**Comment by metaMMA** [comment]
@gazpachoking For my use-case, I don't think so. 

If people wanted only specific keys in the .json file, then... maybe? I don't know much about json dumping other than you can dump the entire contents.

**Comment by gazpachoking** [comment]
I'd say, for making a json output let's skip templates and use a json dump. If we need custom templates, let's maybe think about renaming make_html, which already outputs based on a jinja template (and actually is not html specific at all except for the included templates.) We could also include more templates if need be.

**Comment by metaMMA** [comment]
For example, if I wanted to iterate over all (not necessarily known ahead of time) keys in the .json file, but I want to exclude `original_title`, `original_url`, `accepted_by` `quality`, `task`.

Hmm, I guess my post-processor could do that too.

File: flexget/plugins/output/json.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+from __future__ import unicode_literals, division, absolute_import
+from builtins import *  # noqa pylint: disable=unused-import, redefined-builtin
+
+import os
+import json
```
Comment (gazpachoking) [inline flexget/plugins/output/json.py]: If we use `flexget.utils.json` we'll get date and datetime encoding as well.

File: flexget/plugins/output/json.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        if not config.get('template'):
+            config['template'] = 'json.template'
+
+        filename = os.path.expanduser(config['template'])
+        output = os.path.expanduser(config['file'])
+        # Output to config directory if absolute path has not been specified
+        if not os.path.isabs(output):
+            output = os.path.join(task.manager.config_base, output)
+
+        json.dump(output)
```
Comment (gazpachoking) [inline flexget/plugins/output/json.py]: We need to pass the filename, as well as the object to encode here, which'll be something like `task.accepted`. We might have to do some work on it first to strip out anything that isn't json serializable.

Comment (metaMMA) [reply flexget/plugins/output/json.py]: I used much of the code from lines 85+ in the `db.py` as a `serialize` method in `entry.py`. I'm not sure if this is what you had in mind. But this pull-request relies on that method. I can move it here, if needed, though.

**Comment by metaMMA** [comment]
We also might consider consolidating all of the `make` outputs.
output_options = `html or rss or json or csv or txt or future file output`

```
task:
  <task name>:
    make:
      <output_option>:
        file: <output file name/location>
        encoding: <encoding>
        template: <template filename/location>

**Comment by metaMMA** [comment]
I think this one is ready to go if any of you all could do one last code-review.

File: flexget/plugins/output/json.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+        output = os.path.expanduser(config['file'])
+        # Output to config directory if absolute path has not been specified
+        if not os.path.isabs(output):
+            output = os.path.join(task.manager.config_base, output)
+
+        entries = {}
+        for entry in task.accepted:
+            entry_dict = dict(entry)
+            entries[entry_dict['title']] = entry_dict
```
Comment (paranoidi) [inline flexget/plugins/output/json.py]: Should the plugin be configurable what the format is. What if someone wants to use url or some other field instead? I would have expected the output to be the most straightforward list of resources, ie.

```
[
  {'title': 'asdf1', 'url': 'foobar1'},
  {'title': 'asdf2', 'url': 'foobar2'},
  {'title': 'asdf3', 'url': 'foobar3'}
]
```

**Pull request closed without merging**